pkg/spdk/server.go

```diff
@@ -217,33 +224,85 @@ func (s *Server) ReplicaSnapshotDelete(ctx context.Context, req *spdkrpc.Snapsho
 }
 
 func (s *Server) EngineCreate(ctx context.Context, req *spdkrpc.EngineCreateRequest) (ret *spdkrpc.Engine, err error) {
-	portStart, _, err := s.portAllocator.AllocateRange(1)
-	if err != nil {
-		return nil, err
+	if req.Name == "" || req.VolumeName == "" {
+		return nil, grpcstatus.Error(grpccodes.InvalidArgument, "engine name and volume name are required")
+	}
+	if req.SpecSize == 0 {
+		return nil, grpcstatus.Error(grpccodes.InvalidArgument, "engine spec size is required")
+	}
+	if req.Frontend != types.FrontendSPDKTCPBlockdev && req.Frontend != types.FrontendSPDKTCPNvmf {
+		return nil, grpcstatus.Error(grpccodes.InvalidArgument, "engine frontend is required")
 	}
 
-	bdevAddressMap, err := s.replicaAddressMapToBdevAddressMap(portStart, req.ReplicaAddressMap)
-	if err != nil {
-		return nil, err
+	s.Lock()
+	defer s.Unlock()
+
+	if _, ok := s.engineMap[req.Name]; ok {
+		return nil, grpcstatus.Errorf(grpccodes.AlreadyExists, "engine %v already exists", req.Name)
 	}
 
-	return svcEngineCreate(s.spdkClient, req.Name, req.Frontend, bdevAddressMap, portStart)
+	s.engineMap[req.Name] = NewEngine(req.Name, req.VolumeName, req.Frontend, req.SpecSize)
+	e := s.engineMap[req.Name]
+
+	return e.Create(s.spdkClient, req.ReplicaAddressMap, s.getLocalReplicaBdevMap(req.ReplicaAddressMap), s.portAllocator)
+}
```

Comment:
Should we make sure the `e.Create`successfully  first before assigning to `s.engineMap`?

Reply:
No. Otherwise, later on the server cannot find the failed engine for cleanup.